docs/proposals/server-side-apply-field-managers.md

```diff
@@ -0,0 +1,123 @@
+```
+
+---
+
+## Open Questions
+
+### [Q-1] What should the behavior be for a server-side applied resource upon Application deletion?
+
+The current behavior is to delete all objects "owned" by the Application. A user could choose to leave the resource behind with `Prune=false`, but that would also leave behind any config added to the shared object. Should the default delete behavior for server-side applied resources be to remove any fields that match that Application's [field manager][2]?
+
+### [Q-2] What sync status should the UI display on a shared resource?
+
+If an Application has successfully applied its partial spec to a shared resource, should it display as "in sync"? Or should it show "out of sync" when there are other changes to the shared object?
+
+## Summary
+
+ArgoCD supports server-side apply, but it uses the same field manager, `argocd-controller`, no matter what Application is issuing the sync. Letting users set a field manager, or defaulting to a unique field manager per application would enable users to:
+
+- Manage only specific fields they care about on a shared resource
+- Avoid deleting or overwriting fields that are managed by other Applications
+
+## Motivation
+
+There exist situations where disparate Applications need to add or remove configuration from a shared Kubernetes resource. Server-side apply supports this behavior when different field managers are used.
+
+## Goals
+
+All following goals should be achieve in order to conclude this proposal:
+
+#### [G-1] Applications can apply changes to a shared resource without disturbing existing fields
+
+A common use case of server-side apply is the ability to manage only particular fields on a share Kubernetes resource, while leaving everything else in tact. This requires a unique field manager for each identity that shares the resource.
+
+#### [G-2] Applications that are destroyed only remove the fields they manage from shared resources
+
+A delete operation should undo only the additions or updates it made to a shared resource, unless that resource has `Prune=true`.
+
+#### [G-3] Users can define a field manager as a sync option
+
+Some users may rely on the current behavior emanating from the use of the same field manager across all Applications. The current behavior being that each server side apply sync overwrites all fields on a resource. In other words, "latest sync wins."
+
+## Non-Goals
+
+N/A
+
+## Proposal
+
+1. Add a new sync option named `FieldManager` that accepts a string up to 128 characters in length. This can only be set on an individual resource. Don't allow this sync option to be set on the Application: accidentally overriding the resource-level field manager may have undesirable side effects like leaving orphaned fields behind. When a sync is performed for a resource and server side-apply is enabled, it uses the `FieldManager` if it is set, otherwise it defaults to the hard-coded `argocd-controller` field manager.
```

Comment:
This touches an interesting point about this proposal. The transitioning from standard manager (`argocd-controller`) to dynamic one and vice-versa seems to lead to orphaned fields. This seems to be a problem not only if the new sync option is applied at the application level but also at the resource level. How to address this situation to avoid orphaned fields is something that would have to be clarified as part of this proposal if we want to move forward with the implementation.

Reply:
This could be a case where we want to make that sync option immutable once set. Gonna think this over 🤔

Reply:
Actually, I'm not sure immutability of field managers even feasible with ArgoCD given that:

> Argo CD is largely stateless. All data is persisted as Kubernetes objects, which in turn is stored in Kubernetes' etcd. Redis is only used as a throw-away cache and can be lost.

There would be no way to enforce immutability, as there is no where to keep track of the field manager.

Orphaned fields are a natural consequence of changing field managers. And they can be reclaimed by switching the field manager back to the original. **I'd advocate for just giving users this power for shared resources.** The [ArgoCD docs](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#server-side-apply) already call out that a use case for server-side apply is shared resources, and it doesn't keep track of the last applied state:

> - Patching of existing resources on the cluster that are not fully managed by Argo CD.
> - Use a more declarative approach, which tracks a user's field management, rather than a user's last applied state.

Users taking advantage of server-side apply for shared resources are already bending the Application ownership model. We can simply advise caution.

Reply:
Ok.. Can you please add an orphan demo as well so we can experiment with it and better understand how it will behave?

Reply:
Let's continue the orphaned fields conversation here: https://github.com/argoproj/argo-cd/pull/17125#discussion_r1503358691